tests/test_pytypes.cpp

```diff
@@ -1037,6 +1037,27 @@ TEST_SUBMODULE(pytypes, m) {
 #else
     m.attr("defined_PYBIND11_TEST_PYTYPES_HAS_RANGES") = false;
 #endif
+
+#if defined(PYBIND11_CPP17)
+    m.attr_with_type_hint<py::typing::List<int>>("list_int") = py::list();
+    m.attr_with_type_hint<py::typing::Set<py::str>>("set_str") = py::set();
+
+    struct Empty {};
+    py::class_<Empty>(m, "EmptyAnnotationClass");
+
+    struct Point {
+        float x;
+        py::dict dict_str_int;
+    };
+    auto point = py::class_<Point>(m, "Point");
+    point.attr_with_type_hint<float>("x");
```

Comment:
`x` is not connected to anything on the C++ side.
Could you change the test to how `attr_with_type_hint` would be used in practice?
I am more familiar with `def_readwrite` and it connects like this:
`point.def_readwrite("x", &x)`
and generates getters/setters using `@property`.
How does this connection work with `attr`/`attr_with_type_hint`?

Reply:
```cpp
 struct Point {
        Point(float x): x(x) {}
        float x;
        py::dict dict_str_int;
    };
    auto point = py::class_<Point>(m, "Point");
    point.def(py::init<float>(), "x"_a);
    point.attr_with_type_hint<float>("x");
    point.attr_with_type_hint<py::typing::Dict<py::str, int>>("dict_str_int") = py::dict();
```

```python
def test_class_attribute_types() -> None:
    empty_annotations = get_annotations_helper(m.EmptyAnnotationClass)
    annotations = get_annotations_helper(m.Point)

    assert empty_annotations == {}
    assert annotations["x"] == "float"
    assert annotations["dict_str_int"] == "dict[str, int]"
    p = m.Point(1)
    assert p.x == 1
```

This would fail with
```bash
>       assert p.x == 1
E       AttributeError: 'pybind11_tests.pytypes.Point' object has no attribute 'x'
```

Reply:
I have updated the test case. From what I can tell `attr` is used as a way to modify and access static variables of a class.

```C++
 struct Point {};
    auto point = py::class_<Point>(m, "Point");
    point.def(py::init());
    point.attr_with_type_hint<py::typing::ClassVar<float>>("x");
    point.attr_with_type_hint<py::typing::ClassVar<py::typing::Dict<py::str, int>>>("dict_str_int") = py::dict();
```

```python
def test_class_attribute_types() -> None:
    empty_annotations = get_annotations_helper(m.EmptyAnnotationClass)
    annotations = get_annotations_helper(m.Point)

    assert empty_annotations is None
    assert annotations["x"] == "ClassVar[float]"
    assert annotations["dict_str_int"] == "ClassVar[dict[str, int]]"

    m.Point.x = 1.0
    assert m.Point.x == 1.0

    point = m.Point()
    assert point.x == 1.0

    point.dict_str_int["hi"] = 3
    assert m.Point().dict_str_int == {"hi": 3}
```

Reply:
So far, I have used `.attr` only to access Python bindings on the C++ side.
For the "modify" part, it is used to add additional attributes on the Python side when we do not need any binding to the C++ side?
Then your test makes sense.

Having `ClassVar` and `Final` available makes it possible to add:
* Instance Variables - without any typing class - not static nor const
* Class Variables with `ClassVar` - similar to static
* Final Variables with `Final` - similar to const+static)

Reply:
Just so you are aware `.attr` doesn't seem to support instance variables.